components/brave_today/browser/html_parsing.cc

```diff
@@ -30,8 +31,14 @@ constexpr auto kSupportedRels =
 
 }  // namespace
 
-std::vector<GURL> GetFeedURLsFromHTMLDocument(const std::string& html_body,
+std::vector<GURL> GetFeedURLsFromHTMLDocument(const std::string& charset,
+                                              const std::string& raw_body,
                                               const GURL& html_url) {
+  std::string html_body;
+  if (!base::ConvertToUtf8AndNormalize(raw_body, charset, &html_body)) {
+    return std::vector<GURL>();
+  }
```

Comment:
So...HTML also has (or rather can have) an internal encoding declaration. Would it make sense to use that here like this patch does with RSS?

Of course this doesn't seem to be using a real html parser; it's pulling out links with regular expressions and passing them on if they look like ascii. For the [Ukrainian example](https://www.pravda.com.ua/) in original issue this likely works as well without conversion: the Windows-1251 encoding has the markup punctuation is the same place, and Cyrillic content would be rejected because it has the high bit set same as utf-8. Probably a utf-16 encoding wouldn't work without conversion, however.

In this particular example, the server is sending the encoding in the http header, but only declaring it internally in the rss document. So we'd lose fidelity if we dropped this conversion. Ideally we'd use both the document and http header but of course feed-rs doesn't provide an interface for that.

I just want to make sure we're not tuning for what happens to work with these two documents.

Reply:
Okay, so for what it's worth, I think using a real HTML parser might be the best long term solution here (but maybe this is the extent of the encoding issues, :laughing: ). However, I think this fix is strictly better than what we had before, and is worth landing as is.

1. I think for UTF-16, you have to specify the charset in the `Content-Type` header, as the `meta` tag needs to be ASCII compatible. [1]
2. For ASCII compatible HTML docs, we should still be able to parse the <link> tags, and extract the urls (unless they have non-ascii characters in them).
3. Any documents with the correct charset in the `Content-Type` header should work.
4. Web Server headers take precedence over `meta` attributes [2], so if there's a `charset` in the `Content-Type` header that overrides the `meta` tag anyway.

With the `feed_rs` stuff, I had a little play around and it doesn't seem to handle UTF16 encoded content. With their current API, I think we'd just have to guess, something like:
1. Try passing content as is.
2. Try converting to UTF16. If it works, update the meta tag to be UTF8 (or `feed_rs` will try and convert it again).
3. Repeat, for UTF32. 

[1]: https://www.w3.org/International/questions/qa-html-encoding-declarations#utf16
[2]: https://stackoverflow.com/a/34802103/3260044

Reply:
:facepalm: I didn't even think to check the docs. Okay, that seems like a sensible approach. I wonder if we'll need to replace the `encoding="UTF-16"` attribute?

Reply:
I might defer it to a follow up PR though. I've created https://github.com/brave/brave-browser/issues/26846 so we don't forget :smile: